### PR TITLE
Remove `_manual_slice` constructors

### PR DESCRIPTION
**Pull request by robertbastian**
We added these because range slicing `&foo[x..y]` is not available in const. However, it can be worked around with `slice::split_at`/`slice::split_at_checked`/`slice::split_at_unchecked`. Removing this simplifies our API surface as well as the implementation.

### CODE REVIEW

**Review by sffc — COMMENTED** [review]
We could have done this with `unsafe` originally, but I think we did it with `_manual_slice` in order to avoid `unsafe`. I'll defer to @Manishearth

File: components/locale_core/src/parser/mod.rs
Comment (sffc) [inline components/locale_core/src/parser/mod.rs:54]: `split_at_checked` is new in 1.80: https://github.com/rust-lang/rust/issues/85122

I think I prefer holding this until we can use that in MSRV so we don't introduce `unsafe` code.

Comment (robertbastian) [reply components/locale_core/src/parser/mod.rs:54]: We did this because clients didn't have a safe way to do this at all. I'm assuming the majority of our clients will be on stable, so now the majority of our clients have a safe way to do this. One line more or less unsafe code to make this work isn't the end of the world. Also note that I have a follow-up PR that simplifies the unsafety, but it needs to be benched.

Comment (sffc) [reply components/locale_core/src/parser/mod.rs:54]: ok, if this is just about getting rid of `_manual_slice` in public APIs, then just remove those functions without adding unsafe code?

Comment (robertbastian) [reply components/locale_core/src/parser/mod.rs:54]: The parser calls the `_manual_slice` functions, because it needs to const-slice the input. I'm doing this in one location in the parser, and this is not the only unsafe code in this crate.

**Review by Manishearth — APPROVED** [review]
I'm fine with doing this if we commit to using the safe API later. It's unsafe code being added to a crate (unfortunate), but it's rather straightforward, and const, and only one crate, so I'm not too bothered.

I would like us to move to split_at_checked when we can, though.


(Do we have an issue to keep track of all the rolling "let's fix this with the next MSRV" changes we keep deciding on? I think I saw one before but I don't know if we still have one)

File: components/locale_core/src/parser/mod.rs
Comment (sffc) [reply components/locale_core/src/parser/mod.rs:54]: Please add at least a comment here saying something like

```rust
// TODO(MSRV): Use slice_at_checked in Rust 1.80
```

**Review by sffc — COMMENTED** [review]
OK since @Manishearth approved, but let's record somewhere that we can target this unsafe block after MSRV moves

**Pull request merged**
